extension-manifest-v3/src/background/stats.js

```diff
@@ -95,91 +94,97 @@ const delayMap = new Map();
 function updateIcon(tabId) {
   if (delayMap.has(tabId)) return;
 
-  const timeoutId = setTimeout(
-    () => {
-      delayMap.delete(tabId);
-      refreshIcon(tabId);
-    },
-    // Firefox flickers when updating the icon, so we should expand the throttle
-    __PLATFORM__ === 'firefox' ? 1000 : 250,
+  delayMap.set(
+    tabId,
+    setTimeout(
+      () => {
+        delayMap.delete(tabId);
+        refreshIcon(tabId);
+      },
+      // Firefox flickers when updating the icon, so we should expand the throttle
+      __PLATFORM__ === 'firefox' ? 1000 : 250,
+    ),
   );
 
-  delayMap.set(tabId, timeoutId);
   refreshIcon(tabId);
 }
 
-updateIcon.cancel = (tabId) => {
-  const timeoutId = delayMap.get(tabId);
-  if (timeoutId) {
-    clearTimeout(timeoutId);
-    delayMap.delete(tabId);
-  }
-};
-
-export async function getStatsWithMetadata(since) {
-  const result = await getMergedStats(since);
+export function setupTabStats(tabId, domain) {
+  flushTabStatsToDailyStats(tabId);
 
-  const patternsDetailed = [];
-  for (const key of result.patterns) {
-    const pattern = await trackerDb.getPattern(key);
-    if (pattern) patternsDetailed.push(pattern);
+  if (domain) {
+    tabStats.set(tabId, {
+      domain,
+      trackers: [],
+    });
+  } else {
+    tabStats.delete(tabId);
   }
 
-  return Object.assign(result, { patternsDetailed });
+  updateIcon(tabId);
 }
 
-export async function updateTabStats(tabId, requests) {
-  const stats = tabStats.get(tabId);
-
-  // Stats might not be available on Firefox using webRequest.onBeforeRequest
-  // as some of the requests are fired before the tab is created, tabId -1
-  if (!stats) return;
+export function updateTabStats(tabId, requests) {
+  Promise.resolve().then(async () => {
+    const stats = tabStats.get(tabId);
 
-  for (const request of requests) {
-    const pattern = await trackerDb.getMetadata(request, {
-      getDomainMetadata: true,
-    });
+    // Stats might not be available on Firefox using webRequest.onBeforeRequest
+    // as some of the requests are fired before the tab is created, tabId -1
+    if (!stats) return;
 
-    if (pattern) {
-      let tracker = stats.trackers.find((t) => t.id === pattern.id);
+    let sortingRequired = false;
 
-      if (!tracker) {
-        tracker = { ...pattern, requests: [] };
-        stats.trackers.push(tracker);
-      }
+    // Filter out requests that are not related to the current domain
+    // (e.g. requests on trailing edge when navigation to a new page is in progress)
+    requests = requests.filter(
```

Comment:
This protects from leaking requests from another domain, so the current one is different (using `initiator` property from request details).